### DISCUSSION TITLE
Transmission UI mockup for Linux desktops

### DISCUSSION THREAD

**Discussion by gitSaptarshi** [transmission/transmission#3719] [Ideas]
Transmission is getting pretty dated in terms of UI in Linux. I am proposing a mockup based on recent GTK styling. This mockup makes it look pretty appealing to users and also updates the UI. Here's the mockup I am proposing:
![mockup-transmission](https://user-images.githubusercontent.com/37685073/117259553-c84aa180-ae6b-11eb-91aa-31e6e2e59876.png)
Github doesn't let me upload the SVG file, instead I am including a [mockup-transmission.zip](https://github.com/transmission/transmission/files/6432691/mockup-transmission.zip) file of the design.

**Comment by ckerr** [comment]
The GTK client needs so much love at this point that it hurts to look at the code. 100% agree that a refresh would be good for the project.

Feedback on this version of the mockup --
- I like the menu reorg. 
- I like the layout of the action buttons on the torrents themselves. This is very attractive!
- Overall this shows a lot of promise and I'd like to see another iteration of it

Some questions / issues that could be resolved:
- I'm unsure about the filtering changes. Filtering by state has been retained, but filtering by tracker and by substring has been removed. To me the real power of the current filterbar comes from the combination of these three filters, which is lost here.
- Counts have been removed from the filter. This isn't _as_ important as the filtering functionality itself, but the current implementation's filter dropdown is a nice, terse way of seeing how many torrents you have for each state and each tracker.
- This iteration loses the ability to perform actions on a selected set of torrents, since the current "Torrents" menu has been moved into a per-torrent drop-down. Sample use case: filter by seeding + tracker name -> select all -> pause. This is a pretty common use case when permaseeding and this draft is not able to filter nor perform batch actions. 
- Compact mode seems to have been removed? This is a real problem for permaseeding, as it's not uncommon to have thousands of torrents. I'd be very interested to see what a GNOME refresh of compact mode would look like.
- From this screenshot I'm unsure of what will be in the sidebar.

Thank you for this draft! I'm very encouraged by this work and hope that you continue to revise it!

**Comment by wdlkmpx** [comment]
The GTK2 UI still looks more appealing, a utorrent GTK clone would look even better.
The problem is that the gtk API changes so wildly that applications are not upgradable to the next major version, the mobile UI becomes and more prevalent, the desktop is dying a horrible death thanks to GNOME.

**Comment by illdeletethis** [comment]
sorry to say, but i think your suggestion has all the worst aspects of current web mobile and sadly now also desktop design.
elements in the list take up excessive space, with obscene amounts of unused space in between.
commonly used settings are hidden in submenus of menus, there is too much focus on looking clean over being usable.
your mockup already demonstrates the downsides of your design by having to hide everything exceeding 4 torrents, once you get into the double digits it becomes near unusable.
the current style may be plain, but it works well.
one thing i do like a lot about your concept is moving the start/pause to the individual torrents, layering this onto the folder/file icon of the current GTK interface would be a welcome change.

**Comment by bitigchi** [comment]
It looks better than the previous proposal, which was submitted some years ago. It is a pretty mock-up for GTK standards, but however those standards aren't so high to begin with compared to other platforms.

**Comment by killemov** [comment]
Again a design with a lot of wasted screen real-estate. The progress-bar seems to be the common element to most of these designs. It alone takes up a lot of space to display a percentage. I am against the duplication of control buttons for each torrent. Keeping the main menus opening from the top is an improvement. The separation of torrents with different status (downloading, paused, ...) also doesn't work on a single page because of sorting. Maybe try just sorting or filtering by status.

I assume the current web-UI still uses those menus opening from the bottom AND still use right-mouse-button-activation?

And now for something completely different:
https://forum.transmissionbt.com/viewtopic.php?f=8&t=12555 
https://github.com/killemov/Shift

**Comment by belenos** [comment]
I've been using Fragments lately. It follows Gnome's UI guidelines and looks native. Transmission UI at this point is just visually outdated. I'd love to use it again, it something changes.

**Comment by H2L29** [comment]
please consider the transmission-remote-gtk GUI as a better alternative, clearer, more complete and close to most other torrent clients, I would very much like to see the same design on transmission-gtk

**Comment by bertini97** [comment]
Please, don't forget to consider people that like good-ol-menubars and don't use gnome!

I understand that since gnome is pushing CSD to all their core apps, in their desktop an old style app looks different, but the same is true in reverse. Gnomifying transmission will make it different in other desktops. Adding to this, I personally find the menubar style applications way more usable, and hamburger menus too simplified; I'm not saying one is better than the other, I'm just reminding you people who prefer menubars exist :).

Please, read also consider [this discussion about CSD in Libreoffice's list](https://bugs.documentfoundation.org/show_bug.cgi?id=115512), and in particular:
> Personally, I don't like the CSD design. I find it much easier to select and move windows around and also to tell which document is open with a dedicated title bar. Screen space is not a problem for me. Therefore, if this feature is ever implemented, I request that it be optional.

I, and i suppose some other non-gnome users, ask the same.

Big thanks to anyone that contributes!

**Comment by Pentaphon** [comment]
> The GTK client needs so much love at this point that it hurts to look at the code. 100% agree that a refresh would be good for the project.

Maybe it's time to drop the GTK client altogether since nobody seems interested in actually fixing the GTK client and the Qt client is so far ahead and works on both Windows and Linux.

There's a dev who made a GTK4 frontend called Fragments which uses Transmission as their backend so if Linux users really want to keep using GTK, they can just use that client and let the main project be Qt-only from now on.

https://gitlab.gnome.org/World/Fragments

**Comment by belenos** [comment]
> There's a dev who made a GTK4 frontend called Fragments which uses Transmission as their backend so if Linux users really want to keep using GTK, they can just use that client and let the main project be Qt-only from now only.
> 
> https://gitlab.gnome.org/World/Fragments

I've been using Fragments for a couple of months and loving it so far. Feels more integrated to the system.

**Comment by JaneSmith** [comment]
Fragments fits the GNOME HIG, but it's an extremely simplified client. I personally want the power of Transmission, together with the clean and beautiful HIG of GNOME. The mock-ups in the OP look great to me, and would be clearly more featured than Fragments with e.g. the sidebar categories.

**Comment by Pentaphon** [comment]
> Fragments fits the GNOME HIG, but it's an extremely simplified client. I personally want the power of Transmission, together with the clean and beautiful HIG of GNOME. The mock-ups in the OP look great to me, and would be clearly more featured than Fragments with e.g. the sidebar categories.

Then you need to vote for this issue and get somebody to do a PR for it. 

https://github.com/transmission/transmission/issues/3038

**Comment by GaryElshaw** [comment]
This is a really interesting topic for discussion. Generally, i think it's terrible if a third-party is doing a better job of the ui than you are.

**Reply by Pentaphon** [reply]
> This is a really interesting topic for discussion. Generally, i think it's terrible if a third-party is doing a better job of the ui than you are.

Yeah but the sad fact is that Fragments has literally fragmented the userbase by forcing people who love modern GTK to have to switch to another client instead of the dev just coming to this project and working on the GTK port so that all the users can stay on one client. 

I think the intent is to make a "Gnome Torrent client" so it can be included in the "Gnome Circle" aka https://circle.gnome.org/ but that's not a good enough excuse to fragment the userbase IMHO. I hope the Fragments dev can be convinced to merge his project into Transmission but I don't think that is likely.

**Reply by belenos** [reply]
Instead of accusing Fragments' dev of anything, we should attend to the fact that Transmission has 160+ contributors and none of them have showed real interest on porting it to GTK4 and/or GNOME HIG. Gnome 3 was launched in 2011. It has been too long to ignore it.

**Reply by Pentaphon** [reply]
> Instead of accusing Fragments' dev of anything

Have you asked them why they created Fragments instead of updating the GTK port of Transmission?

**Reply by belenos** [reply]
You are new here, aren't you? 😂

**Reply by Pentaphon** [reply]
Are you going to answer the question because I don't know what you're talking about?

**Reply by wdlkmpx** [reply]
> Are you going to answer the question because I don't know what you're talking about?

I don't know either but I remember this:

https://trac.transmissionbt.com/ticket/3685

> I guess you have to decide if you are a GNOME app, an Ubuntu app, or an XFCE app unfortunately. I'm sorry that this is the case but it wasn't GNOME's fault that Ubuntu has started this fork. And I have no idea what XFCE is or does sorry.

> It is my hope that you are a GNOME app. Yes this kind of fragmentation is unfortunate. I'm not happy about it either. Anyway, I just wanted to give you a heads up. Wish you the best.

And this:
https://linuxreviews.org/GNOME_Videos

If you want your app to have features and options, it obviously can't be a gnome app, that may be the answer

**Comment by ckerr** [comment]
> Instead of accusing Fragments' dev of anything,

:+1: Everyone can and should work on whatever they want. The more the merrier.

> none of them have showed real interest on porting it to GTK4 and/or GNOME HIG

Be the change you want to see...

**Comment by mikedld** [comment]
> Transmission has 160+ contributors

I doubt it, at last if we count active contributors.

> real interest on porting it to GTK4 and/or GNOME HIG

As vaguely implied in the Trac ticket (linked above), Transmission is a GTK app, not a GNOME app. And IMHO one shouldn't equate GTK to GNOME because the latter is just one DE and the former is used by quite a few other DEs (and not to mention non-DE environments using GTK-based apps) that don't care about GNOME design principles that much. In recent years GNOME somehow managed to influence GTK [in a bad way] to drop functionality and even whole widgets that aren't up to their HIG, which is crazy: if you don't want to use some functionality or widget in GNOME, just don't use it, but don't force people to rewrite their apps (in a backwards-incompatible way) every few years, reimplementing the functionality they need (if at all possible) that was previously provided by the toolkit. If the app then doesn't appear to match your standard, well boohoo, don't include it in "standard GNOME package" (whatever that might be) and don't grant it "GNOME design award" (if it's a thing).

I used to love GTK a bit more than Qt (for a few reasons), but GTK people should really learn backward compatibility from the other toolkits out there. And if they claim that they "don't have enough manpower" for it, well then maybe they're doing something wrong with the project to begin with that it doesn't attract contributors (yep, sounds pretty much like an egg/chicken problem, doesn't make the point less valid).

According to #3689, people *are* interested in Transmission having GTK client (although the final call to drop it or not is on those who actually write code, so I can't say I like that poll). Where there's demand, supply will follow. Just be patient.

**Reply by Pentaphon** [reply]
> According to #3689, people _are_ interested in Transmission having GTK client (although the final call to drop it or not is on those who actually write code, so I can't say I like that poll). Where there's demand, supply will follow. Just be patient.

The whole demand/supply curve depends wholly on if there is profit to be made, and on an open source project, profit isn't even part of the equation in the overwhelming majority of cases so there's literally no incentive to develop other than enthusiasm or a need to develop one's skillset. My poll shows that there is indeed interest in having the GTK port updated, but without GTK-enthusiastic or self-taught GTK developers, there's no hope of any development ever happening. 

2 reasons why I suggested dropping GTK in favor of Qt:  

1) By shifting the focus to Qt, we may end up attracting more developers, as there's absolutely more Qt-experienced developers than there are GTK-experienced developers out there in the professional and self-taught world. 

2) The Linux and Windows ports both benefit from Qt development so 2 birds with 1 stone, etc. 

Don't get me wrong, I would love if Transmission would get more GTK development, but it's just not there anymore.

**Comment by linsui** [comment]
> There's a dev who made a GTK4 frontend called Fragments which uses Transmission as their backend so if Linux users really want to keep using GTK, they can just use that client and let the main project be Qt-only from now on.

> 2 reasons why I suggested dropping GTK in favor of Qt:

I have a crazy idea... How about dropping all of QT, GTK and whatever used for MacOS and let the main project be daemon with a web ui only? There's absolutely more JS-experienced developers than there are QT/GTK/MacOS-experienced developers out there in the professional and self-taught world. The Linux and Windows and MacOS/Android/iOS/BSD/whatever ports both benefit from JS development so many birds with 1 stone, etc. Syncthing is a good example.

Anyone who wants a desktop client can write third-party clients with what they want to use. There is already Fragments and there can also be a third-party QT client since there are so much more QT devs than GTK devs.

**Reply by Pentaphon** [reply]
> I have a crazy idea... How about dropping all of QT, GTK and whatever used for MacOS and let the main project be daemon with a web ui only?

Probably because most people want a native UI.

**Reply by killemov** [reply]
> > I have a crazy idea... How about dropping all of QT, GTK and whatever used for MacOS and let the main project be daemon with a web ui only?

I have advocated for dropping all native UI's as well, only continuing development on daemon, CLI and Web-UI. This focuses development on these domains only. 

> > Probably because most people want a native UI.

I would like to see a poll for that. A native UI usually has an "I can do more with it" sentiment attached to it. If that, being able to do more with the native UI, is the case for Transmission then that should be remedied. Today I would only write a native UI for software that is graphics intensive anyway period.

**Reply by linsui** [reply]
It's not necessary to drop the native UI. Since there are so many people like native UI, they can develop a third-party client. It's easy to just wrap the webui with Electron or Tauri or whatever. QT fans can pick the QT code. Then QT fans and GTK fans don't need to argue here anymore. They can talk with their code. And the Transmission devs can focus on the daemon the improvment can benefit everyone.
